libtbl.3
.Dd $Mdocdate$
.Dt LIBTBL 3
.Os
.Sh NAME
.Nm libtbl_tbl_init , libtbl_tbl_add_row , libtbl_tbl_end
.Nd Generic Table Printer
.Sh LIBRARY
.Lb libtbl
.Sh SYNOPSIS
.In libtbl.h
.Ft libtbl_tbl
.Fn "libtbl_tbl_begin" "libtbl_tblcoldef cols[]" "size_t cols_size"
.Ft int
.Fn libtbl_tbl_add_row "libtbl_tbl table" "..."
.Ft void
.Fn libtbl_tbl_end "libtbl_tbl table"
.Ft int
.Fn libtbl_set_colours_enabled "int enabled"
.Sh DESCRIPTION
libtbl is a library for printing tabular data. Given a set of columns
defined by the following data structure it will accept a list of
rows and print them out.
.Bd -literal -offset indent
typedef struct libtbl_tblcoldef libtbl_tblcoldef;

struct libtbl_tblcoldef {
	char const *const name;     /* header of the column */
	int flags;                  /* display flags */
	int type;                   /* data type of the column */
};
.Ed
.Pp
The name field will be printed above the first row of data as is.
The flags field is a bit mask of the following values:
.Bl -tag -width LIBTBL_TBLCOL_COLOUREXPL
.It Dv LIBTBL_TBLCOL_JUSTIFYR
Justify the contents of this column to the right.
.It Dv LIBTBL_TBLCOL_BOLD
Print the column contents with a bold font.
.It Dv LIBTBL_TBLCOL_256COLOUR
Accept an additional uint64_t containing the RGBA colour to print the
cell contents in.
.It Dv LIBTBL_TBLCOL_CUSTOM
Accept a callback to do custom formatting. See
.Sx CUSTOM FORMATTERS
.It Dv LIBTBL_TBLCOL_COLOUREXPL
Accept an additional colour code for the cell contents. The colour
codes are:
.Bl -bullet -compact
.It
.Dv LIBTBL_COLOUR_BLACK
.It
.Dv LIBTBL_COLOUR_RED
.It
.Dv LIBTBL_COLOUR_GREEN
.It
.Dv LIBTBL_COLOUR_YELLOW
.It
.Dv LIBTBL_COLOUR_BLUE
.It
.Dv LIBTBL_COLOUR_MAGENTA
.It
.Dv LIBTBL_COLOUR_CYAN
.It
.Dv LIBTBL_COLOUR_WHITE
.It
.Dv LIBTBL_COLOUR_DEFAULT
.El
.El
.Pp
You can add rows to an initialized table by repeatedly calling
.Fn libtbl_tbl_add_row
and supplying arguments as needed.
When a column is given one of the colour options, provide the
colour first and then the data. When the custom flag is set,
provide a function pointer first, a user-data pointer second
and the cell content last. See
.Sx CUSTOM FORMATTERS
and
.Sx EXAMPLES
for more details.
.Pp
Once you have called
.Fn libtbl_tbl_add_row
to add all the rows, call
.Fn libtbl_tbl_end
to dump the table to stdout and free all the internally allocated
resources.
.Pp
Note that all data passed to the functions as pointers need to be
valid up to this call. An exception to this rule are strings as
they will be copied internally. The
.Fn libtbl_tbl_end
call will also call into your custom callbacks and pass the user pointer
as given.
.Sh CUSTOM FORMATTERS
Whenever
.Dv LIBTBL_TBLCOL_CUSTOM
is specified in a column definition you need to provide a function
pointer, a user data pointer and the cell contents to
.Fn libtbl_tbl_add_row
for a cell of the given column:
.Bd -literal -offset indent
typedef void (*custom_formatter)(int is_start, void *userdata);
.Ed
.Pp
When the cell is about to be printed the function is called with 1 passed to
.Va is_start
and the userdata pointer passed as it was given to the add_row call.
After the contents have been printed your callback will be called
again with a 0 instead of a 1 to indicate that you should stop
formatting the output.
.Pp
See
.Pa example.c
for an example of how to pass arguments correctly.
.Sh RETURN VALUES
.Fn libtbl_tbl_begin
returns a valid table handle on success or NULL on failure. Both
.Fn libtbl_tbl_add_row
and
.Fn libtbl_tbl_end
return 0 on success or -1 on failure.
.Fn libtbl_set_colours_enabled
returns 0 if colours have been disabled and 1 if colours have been
enabled.
.Sh ENVIRONMENT
.Bl -tag -width LIBTBL_NO_COLOURS
.It Ev LIBTBL_NO_COLOURS
If set to 1 this disables colour output and custom formatters
entirely.
.Sh EXAMPLES
To print a simple table with 3 columns of data:
.Bd -literal -offset indent
#include <libtbl.h>
#include <stdbool.h>

...

libtbl_tbl table;
libtbl_tblcoldef cols[] = {
	{ .name = "COL1", .type = LIBTBL_TBLCOLTYPE_INT, .flags = 0 },
	{ .name = "COL2", .type = LIBTBL_TBLCOLTYPE_BOOL, .flags = 0 },
	{ .name = "COL3", .type = LIBTBL_TBLCOLTYPE_STRING, .flags = 0 },
};

table = libtbl_tbl_begin(cols, sizeof(cols) / sizeof(*cols));
if (!table) {
	fprintf(stderr, "error: could not init table");
	abort();
}

for (...) {
	libtbl_tbl_add_row(table, i, true, "test");
}

libtbl_tbl_end(table);
.Ed
.Pp
You can find more examples shipped in the distribution of libtbl.
.Sh ERRORS
The functions return error codes only if internal resource allocations
failed.
.Sh SEE ALSO
.Xr column 1
.Sh HISTORY
Libtbl was written in 2022 as part of the gcli program and later
put into a separate library as people considered it useful for other
projects.
.Sh AUTHORS
libtbl was written by
.An Nico Sonack Aq Mt nsonack@herrhotzenplotz.de
.Sh BUGS
Currently it is impossible to print to other files than stdout.


libtbl.c
/*
 * Copyright 2022 Nico Sonack <nsonack@herrhotzenplotz.de>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following
 * disclaimer in the documentation and/or other materials provided
 * with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "libtbl.h"

#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

/*****************************************************************************
 * Helpers
 ****************************************************************************/
static char *
libtbl_asprintf(char const *fmt, ...)
{
	/* Obviously this is not ideal but it works */
	char tmp = 0, *result = NULL;
	size_t actual = 0;
	va_list vp;

	va_start(vp, fmt);
	actual = vsnprintf(&tmp, 1, fmt, vp);
	va_end(vp);

	result = calloc(1, actual + 1);
	if (!result)
		return NULL;
 
	va_start(vp, fmt);
	vsnprintf(result, actual + 1, fmt, vp);
	va_end(vp);

	return result;
}

/*****************************************************************************
 * ANSI Colour handling 
 ****************************************************************************/
static int have_colours = -1;

static int
libtbl_have_colours(void)
{
	if (have_colours < 0) {
		have_colours = isatty(STDOUT_FILENO);

		/* allow the environment to disable colours */
		if (have_colours) {
			char const *env_disable_colours = getenv("LIBTBL_NO_COLOURS");

			if (env_disable_colours)
				have_colours = strcmp("1", env_disable_colours);
		}
	}

	return have_colours;
}

static struct {
        uint32_t code;
        char *sequence;
} colour_table[1024];
static size_t colour_table_size;

static void
clean_colour_table(void)
{
        for (size_t i = 0; i < colour_table_size; ++i)
                free(colour_table[i].sequence);
}

static char const *
colour_cache_lookup(uint32_t const code)
{
        for (size_t i = 0; i < colour_table_size; ++i) {
                if (colour_table[i].code == code)
                        return colour_table[i].sequence;
        }
        return NULL;
}

static void
colour_cache_insert(uint32_t const code, char *sequence)
{
        colour_table[colour_table_size].code = code;
        colour_table[colour_table_size].sequence = sequence;
        colour_table_size++;
}

static char const *const
libtbl_setcolour(int code)
{
	if (!libtbl_have_colours())
		return "";

	switch (code) {
	case LIBTBL_COLOUR_BLACK:   return "\033[30m";
	case LIBTBL_COLOUR_RED:     return "\033[31m";
	case LIBTBL_COLOUR_GREEN:   return "\033[32m";
	case LIBTBL_COLOUR_YELLOW:  return "\033[33m";
	case LIBTBL_COLOUR_BLUE:    return "\033[34m";
	case LIBTBL_COLOUR_MAGENTA: return "\033[35m";
	case LIBTBL_COLOUR_CYAN:    return "\033[36m";
	case LIBTBL_COLOUR_WHITE:   return "\033[37m";
	case LIBTBL_COLOUR_DEFAULT: return "\033[39m";
	default:
		return NULL;
	}
	return NULL;
}

static char const *const
libtbl_setcolour256(uint64_t const code)
{
	char *result = NULL;
	char const *oldresult = NULL;

        if (!libtbl_have_colours())
                return "";

	if (colour_table_size == 0)
		atexit(clean_colour_table);

	oldresult = colour_cache_lookup(code);
	if (oldresult)
		return oldresult;

        /* TODO: This is inherently screwed */
	result = libtbl_asprintf("\033[38;2;%02d;%02d;%02dm",
	                         (code & 0xFF000000) >> 24,
	                         (code & 0x00FF0000) >> 16,
	                         (code & 0x0000FF00) >>  8);

	colour_cache_insert(code, result);

	return result;
}

static char const *const
libtbl_setbold(void)
{
	if (!libtbl_have_colours())
		return "";
	else
		return "\033[1m";
}

static char const *const
libtbl_resetcolour(void)
{
	if (!libtbl_have_colours())
		return "";
	else
		return "\033[m";
}

static char const *const
libtbl_resetbold(void)
{
	if (!libtbl_have_colours())
		return "";
	else
		return "\033[22m";
}

int
libtbl_set_colours_enabled(int enabled)
{
	return (have_colours = enabled);
}

/*****************************************************************************
 * Table State
 ****************************************************************************/
/* A row */
struct libtbl_tblrow;

/* Internal state of a table printer. We return a handle to it in
 * gcli_table_init. */
struct libtbl_tbl {
	libtbl_tblcoldef const *cols; /* user provided column definitons */
	int *col_widths;        /* minimum width of the columns */
	size_t cols_size;       /* size of above arrays */

	struct libtbl_tblrow *rows; /* list of rows */
	size_t rows_size;       /* number of rows */
};

struct libtbl_tblrow {
	struct {
		char *text;     /* the text in the cell */
		char const *colour;     /* colour (ansi escape sequence) if
		                         * explicit fixed colour was given */

		custom_formatter formatter_fn;
		void *userdata;         /* user data to pass to the function */
	} *cells;
};

/* Push a row into the table state */
static int
table_pushrow(struct libtbl_tbl *const table, struct libtbl_tblrow row)
{
	table->rows = realloc(table->rows, sizeof(*table->rows) * (table->rows_size + 1));
	if (!table->rows)
		return -1;

	table->rows[table->rows_size++] = row;

	return 0;
}

/** Initialize the internal state structure of the table printer. */
libtbl_tbl
libtbl_tbl_begin(libtbl_tblcoldef const *const cols, size_t const cols_size)
{
	struct libtbl_tbl *tbl;

	/* Allocate the structure and fill in the handle */
	tbl = calloc(sizeof(*tbl), 1);
	if (!tbl)
		return NULL;

	/* Reserve memory for the column sizes */
	tbl->col_widths = calloc(sizeof(*tbl->col_widths), cols_size);
	if (!tbl->col_widths) {
		free(tbl);
		return NULL;
	}

	/* Store the list of columns */
	tbl->cols = cols;
	tbl->cols_size = cols_size;

	/* Check the headers */
	for (size_t i = 0; i < cols_size; ++i) {

		/* Compute the header's length and use these as initial
		 * values */
		tbl->col_widths[i] = strlen(cols[i].name);
	}

	return tbl;
}

static void
table_freerow(struct libtbl_tblrow *row, size_t const cols)
{
	for (size_t i = 0; i < cols; ++i)
		free(row->cells[i].text);

	free(row->cells);
	row->cells = NULL;
}

static int
tablerow_add_cell(struct libtbl_tbl *const table,
                  struct libtbl_tblrow *const row,
                  size_t const col, va_list vp)
{
	int cell_size = 0;

	/* Extract the explicit colour code */
	if (table->cols[col].flags & LIBTBL_TBLCOL_COLOUREXPL) {
		int code = va_arg(vp, int);

		/* don't free that! it's allocated and free'ed up in the colour
		 * handling code. */
		row->cells[col].colour = libtbl_setcolour(code);
	} else if (table->cols[col].flags & LIBTBL_TBLCOL_256COLOUR) {
		uint64_t hexcode = va_arg(vp, uint64_t);

		/* see comment above */
		row->cells[col].colour = libtbl_setcolour256(hexcode);
	} else if (table->cols[col].flags & LIBTBL_TBLCOL_CUSTOM) {
		row->cells[col].formatter_fn = va_arg(vp, custom_formatter);
		row->cells[col].userdata = va_arg(vp, void *);
	}


	/* Process the content */
	switch (table->cols[col].type) {
	case LIBTBL_TBLCOLTYPE_INT: {
		row->cells[col].text = libtbl_asprintf("%d", va_arg(vp, int));
		cell_size = strlen(row->cells[col].text);
	} break;
	case LIBTBL_TBLCOLTYPE_LONG: {
		row->cells[col].text = libtbl_asprintf("%ld", va_arg(vp, long));
		cell_size = strlen(row->cells[col].text);
	} break;
	case LIBTBL_TBLCOLTYPE_STRING: {
		char *it = va_arg(vp, char *);
		row->cells[col].text = strdup(it);
		cell_size = strlen(it);
	} break;
	case LIBTBL_TBLCOLTYPE_DOUBLE: {
		row->cells[col].text = libtbl_asprintf("%lf", va_arg(vp, double));
		cell_size = strlen(row->cells[col].text);
	} break;
	case LIBTBL_TBLCOLTYPE_BOOL: {
		/* Do not use real _Bool type as it triggers a compiler bug in
		 * LLVM clang 13 */
		int val = va_arg(vp, int);
		if (val) {
			row->cells[col].text = strdup("yes");
			cell_size = 3;
		} else {
			row->cells[col].text = strdup("no");
			cell_size = 2;
		}
	} break;
	default:
		return -1;
	}

	/* Update the column width if needed */
	if (table->col_widths[col] < cell_size)
		table->col_widths[col] = cell_size;

	return 0;
}

int
libtbl_tbl_add_row(libtbl_tbl _table, ...)
{
	va_list vp;
	struct libtbl_tblrow row = {0};
	struct libtbl_tbl *table = (struct libtbl_tbl *)(_table);

	/* reserve array of cells */
	row.cells = calloc(sizeof(*row.cells), table->cols_size);
	if (!row.cells)
		return -1;

	va_start(vp, _table);

	/* Step through all the columns and print the cells */
	for (size_t i = 0; i < table->cols_size; ++i) {
		if (tablerow_add_cell(table, &row, i, vp) < 0) {
			table_freerow(&row, table->cols_size);
			va_end(vp);
			return -1;
		}
	}

	va_end(vp);

	/* Push the row into the table */
	if (table_pushrow(table, row) < 0) {
		table_freerow(&row, table->cols_size);
		return -1;
	}

	return 0;
}

static void
pad(size_t const n)
{
	for (size_t p = 0; p < n; ++p)
		putchar(' ');
}

static void
dump_row(struct libtbl_tbl const *const table, size_t const i)
{
	struct libtbl_tblrow const *const row = &table->rows[i];

	for (size_t col = 0; col < table->cols_size; ++col) {
		/* If right justified and not last column, print padding */
		if ((table->cols[col].flags & LIBTBL_TBLCOL_JUSTIFYR) &&
		    (col + 1) < table->cols_size)
			pad(table->col_widths[col] - strlen(row->cells[col].text));

		/* Colour Processing */
		if (table->cols[col].flags &
		         (LIBTBL_TBLCOL_COLOUREXPL|LIBTBL_TBLCOL_256COLOUR))
			printf("%s", row->cells[col].colour);

		/* Bold */
		if (table->cols[col].flags & LIBTBL_TBLCOL_BOLD)
			printf("%s", libtbl_setbold());

		/* Custom formatter handling */
		if (libtbl_have_colours() && (table->cols[col].flags & LIBTBL_TBLCOL_CUSTOM))
			row->cells[col].formatter_fn(1, row->cells[col].userdata);

		/* Print cell if it is not NULL, otherwise indicate it by
		 * printing <empty> */
		printf("%s  ", row->cells[col].text ? row->cells[col].text : "<empty>");

		/* Custom formatter end */
		if (libtbl_have_colours() && (table->cols[col].flags & LIBTBL_TBLCOL_CUSTOM))
			row->cells[col].formatter_fn(0, row->cells[col].userdata);

		/* End colour */
		if (table->cols[col].flags & (LIBTBL_TBLCOL_COLOUREXPL|LIBTBL_TBLCOL_256COLOUR))
			printf("%s", libtbl_resetcolour());

		/* Stop printing in bold */
		if (table->cols[col].flags & LIBTBL_TBLCOL_BOLD)
			printf("%s", libtbl_resetbold());

		/* If left-justified and not last column, print padding */
		if (!(table->cols[col].flags & LIBTBL_TBLCOL_JUSTIFYR) &&
		    (col + 1) < table->cols_size)
			pad(table->col_widths[col] - strlen(row->cells[col].text));
	}
	putchar('\n');
}

static int
libtbl_tbl_dump(libtbl_tbl const _table)
{
	struct libtbl_tbl const *const table = (struct libtbl_tbl const *const)_table;

	for (size_t i = 0; i < table->cols_size; ++i) {
		printf("%s  ", table->cols[i].name);

		if ((i + 1) < table->cols_size)
			pad(table->col_widths[i] - strlen(table->cols[i].name));
	}
	printf("\n");

	for (size_t i = 0; i < table->rows_size; ++i) {
		dump_row(table, i);
	}

	return 0;
}

static void
libtbl_tbl_free(libtbl_tbl _table)
{
	struct libtbl_tbl *tbl = (struct libtbl_tbl *)_table;

	for (size_t row = 0; row < tbl->rows_size; ++row)
		table_freerow(&tbl->rows[row], tbl->cols_size);

	free(tbl->rows);
	free(tbl->col_widths);
	free(tbl);
}

void
libtbl_tbl_end(libtbl_tbl tbl)
{
	libtbl_tbl_dump(tbl);
	libtbl_tbl_free(tbl);
}
